src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -54,17 +54,35 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
     workInProgress.pendingWorkPriority = NoWork;
   }
 
-  function updateClassComponent(current, workInProgress) {
+  function updateClassComponent(current : ?Fiber, workInProgress : Fiber) {
     var props = workInProgress.pendingProps;
-    if (!workInProgress.stateNode) {
+    var instance = workInProgress.stateNode;
+    if (!instance) {
       var ctor = workInProgress.type;
-      workInProgress.stateNode = new ctor(props);
+      workInProgress.stateNode = instance = new ctor(props);
+    } else if (typeof instance.shouldComponentUpdate === 'function') {
+      if (current && current.memoizedProps) {
+        // Revert to the last flushed props, incase we aborted an update.
+        instance.props = current.memoizedProps;
+        if (!instance.shouldComponentUpdate(props)) {
+          return bailoutOnCurrent(current, workInProgress);
+        }
+      }
+      if (!workInProgress.hasWorkInProgress && workInProgress.memoizedProps) {
```

Comment:
Initially I only did this with the cheaper equality check. However, I realize now that when applied to `shouldComponentUpdate`, and later `PureComponent`, doing this twice could be potentially expensive when this is simple rerender. It would be better to do this twice only if this is a restart of already processed work, ignoring the ping-pong case which is rare.